File: java/Tutorials/Media/Java/read-audio/src/main/java/com/neurotec/tutorials/ReadAudio.java
package com.neurotec.tutorials;

import java.io.IOException;
import java.util.EnumSet;

import com.neurotec.lang.NThrowable;
import com.neurotec.licensing.NLicense;
import com.neurotec.media.NAudioFormat;
import com.neurotec.media.NMediaFormat;
import com.neurotec.media.NMediaReader;
import com.neurotec.media.NMediaSource;
import com.neurotec.media.NMediaType;
import com.neurotec.media.NVideoFormat;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class ReadAudio {

	private static final String DESCRIPTION = "Demonstrates reading sound from specified filename or url";
	private static final String NAME = "read-audio";
	private static final String VERSION = "9.0.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [source] [bufferCount] <optional: is url>%n", NAME);
		System.out.println();
		System.out.println("\tsource - filename or url sound buffers should be captured from.");
		System.out.println("\tbufferCount - number of sound buffers to capture from specified filename or url.");
		System.out.println("\tis url - specifies that passed source parameter is url (value: 1) or filename (value: 0).");
		System.out.println();
	}

	public static void main(String[] args) {
		final String components = "Media";

		LibraryManager.initLibraryPath();

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 2) {
			usage();
			System.exit(-1);
		}

		NMediaSource mediaSource = null;
		NMediaReader mediaReader = null;

		try {
			if (!NLicense.obtainComponents("/local", 5000, components)) {
				System.err.println("Could not obtain licenses for components: " + components);
				System.exit(-1);
			}

			String uri = args[0];
			boolean isUrl = false;
			int bufferCount = Integer.parseInt(args[1]);
			if (bufferCount <= 0) {
				System.out.println("no sound buffers will be captured as sound buffer count is not specified");
			}

			if (args.length > 2) {
				isUrl = args[2] == "1";
			}

			// create media source
			mediaSource = (isUrl) ? NMediaSource.fromUrl(uri) : NMediaSource.fromFile(uri);
			System.out.format("display name: %s\n", mediaSource.getDisplayName());

			mediaReader = new NMediaReader(mediaSource, EnumSet.of(NMediaType.AUDIO), true);
			readSoundBuffers(mediaReader, bufferCount);
			System.out.println("done");
		} catch (Throwable th) {
			th.printStackTrace();

			int errorCode = -1;

			if (th instanceof NThrowable) {
				errorCode = ((NThrowable)th).getCode();
			}

			System.exit(errorCode);
		} finally {
			try {
				NLicense.releaseComponents(components);
			} catch (IOException e) {
				e.printStackTrace();
			}
			if (mediaSource != null) mediaSource.dispose();
			if (mediaReader != null) mediaReader.dispose();
		}
	}

	private static void readSoundBuffers(NMediaReader mediaReader, int bufferCount) throws Exception {
		NMediaSource mediaSource = mediaReader.getSource();

		System.out.format("media length: %s\n", mediaReader.getLength());

		NMediaFormat[] mediaFormats = mediaSource.getFormats(NMediaType.AUDIO);
		if (mediaFormats == null) {
			System.err.println("formats are not yet available (should be available after media reader is started");
		} else {
			System.out.format("format count: %s\n", mediaFormats.length);
			for (int i = 0; i < mediaFormats.length; i++) {
				System.out.format("[%s] ", i);
				dumpMediaFormat(mediaFormats[i]);
			}
		}

		NMediaFormat currentMediaFormat = mediaSource.getCurrentFormat(NMediaType.AUDIO);
		if (currentMediaFormat != null) {
			System.out.println("current media format:");
			dumpMediaFormat(currentMediaFormat);

			if (mediaFormats != null) {
				System.out.println("set the last supported format (optional) ... ");
				mediaSource.setCurrentFormat(NMediaType.AUDIO, mediaFormats[mediaFormats.length - 1]);
			}
		} else
			System.err.println("current media format is not yet available (will be availble after media reader start)");

		System.out.print("starting capture ... ");
		mediaReader.start();
		System.out.println("capture started");

		try {
			currentMediaFormat = mediaSource.getCurrentFormat(NMediaType.AUDIO);
			if (currentMediaFormat == null) throw new Exception("current media format is not set even after media reader start!");
			System.out.println("capturing with format: ");
			dumpMediaFormat(currentMediaFormat);

			for (int i = 0; i < bufferCount; i++) {

				NMediaReader.ReadResult result = mediaReader.readAudioSample();
				if (result.getSoundBuffer() == null) return; // end of stream

				System.out.format("[%s %s] sample rate: %s, sample length: %s\n", result.getTimeStamp(), result.getDuration(), result.getSoundBuffer().getSampleRate(), result.getSoundBuffer().getLength());
			}
		} finally {
			mediaReader.stop();
		}
	}

	private static void dumpMediaFormat(NMediaFormat mediaFormat) {
		if (mediaFormat == null) throw new NullPointerException("mediaFormat");
		switch (mediaFormat.getMediaType()) {
		case VIDEO:
			NVideoFormat videoFormat = (NVideoFormat) mediaFormat;
			System.out.format("video format .. %sx%s @ %s/%s (interlace: %s, aspect ratio: %s/%s)\n", videoFormat.getWidth(), videoFormat.getHeight(), videoFormat.getFrameRate().numerator, videoFormat.getFrameRate().denominator, videoFormat.getInterlaceMode(), videoFormat.getPixelAspectRatio().numerator, videoFormat.getPixelAspectRatio().denominator);
			break;
		case AUDIO:
			NAudioFormat audioFormat = (NAudioFormat) mediaFormat;
			System.out.format("audio format .. channels: %s, samples/second: %s, bits/channel: %s\n", audioFormat.getChannelCount(), audioFormat.getSampleRate(), audioFormat.getBitsPerChannel());
			break;
		default:
			throw new IllegalArgumentException("unknown media type specified in format!");
		}
	}
}

File: java/Tutorials/Media/C/ReadAudio/ReadAudio.c
#include <TutorialUtils.h>

#ifdef N_MAC_OSX_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("ReadAudio");
const NChar description[] = N_T("Demonstrates reading sound from specified filename or url");
const NChar version[] = N_T("9.0.0.0");
const NChar copyright[] = N_T("Copyright (C) 2011-2017 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [source] [bufferCount] <optional: is url>\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tsource - filename or url sound buffers should be captured from\n"));
	printf(N_T("\tbufferCount - number of sound buffers to capture from specified filename or url\n"));
	printf(N_T("\tis url - specifies that passed source parameter is url (value: 1) or filename (value: 0)\n"));
	printf(N_T("\n"));
	return 1;
}

static NResult ReadSoundBuffers(HNMediaSource hSource, NInt bufferCount);

int main(int argc, NChar **argv)
{
	HNString hDisplayName = NULL;
	NResult result = N_OK;
	NInt bufferCount = 0;
	const NChar * components = N_T("Media");
	NBool available = NFalse;
	HNMediaReader hReader = NULL;
	NChar * szUri = NULL;
	NBool isUrl = NFalse;
	HNMediaSource hSource = NULL;

	OnStart(title, description, version, copyright, argc, argv);

	if (argc < 3)
	{
		OnExit();
		return usage();
	}

	// check the license first
	result = NLicenseObtainComponents(N_T("/local"), N_T("5000"), components, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtainComponents() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("Licenses for %s not available\n"), components);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	szUri = argv[1];
	bufferCount = atoi(argv[2]);
	if (bufferCount == 0)
	{
		printf(N_T("no sound buffers will be read as sound buffer count is not specified"));
	}

	if (argc > 3)
	{
		isUrl = atoi(argv[3]) == 1;
	}

	// create media source
	result = isUrl ? NMediaSourceCreateFromUrl(szUri, 0, &hSource) : NMediaSourceCreateFromFile(szUri, 0, &hSource);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to create media source (result = %d)!\n"), result);
		goto FINALLY;
	}

	// get source settings
	result = NMediaSourceGetDisplayNameN(hSource, &hDisplayName);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get display name (result = %d)!\n"), result);
		goto FINALLY;
	}
	{
		const NChar * szDisplayName;
		result = NStringGetBuffer(hDisplayName, 0, &szDisplayName);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get display name (result = %d)!\n"), result);
			goto FINALLY;
		}

		printf(N_T("display name: %s\n"), szDisplayName);
	}
	result = NStringSet(NULL, &hDisplayName);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to clear display name (result = %d)!\n"), result);
		goto FINALLY;
	}

	result = NMediaReaderCreate(hSource, nmtAudio, NTrue, 0, &hReader);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to create media reader from media source (result = %d)!\n"), result);
		goto FINALLY;
	}

	result = ReadSoundBuffers(hReader, bufferCount);
	if (NFailed(result)) goto FINALLY;
	
	printf(N_T("done\n"));

	result = N_OK;
FINALLY:
	{
		NResult result2;

		result2 = NObjectSet(NULL, &hReader);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NObjectSet(NULL, &hSource);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NStringSet(NULL, &hDisplayName);
		if (NFailed(result2)) PrintErrorMsg(N_T("NStringSet() failed, result = %d\n"), result2);

		result2 = NLicenseReleaseComponents(components);
		if (NFailed(result2)) PrintErrorMsg(N_T("NLicenseReleaseComponents() failed, result = %d\n"), result2);
	}

	OnExit();

	return result;
}

static void DumpMediaFormat(HNMediaFormat hFormat)
{
	NMediaType formatMediaType;
	NResult result;

	if (hFormat == NULL) return;

	result = NMediaFormatGetMediaType(hFormat, &formatMediaType);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media type for specified format (result = %d)!\n"), result);
		return;
	}

	switch (formatMediaType)
	{
	case nmtVideo:
		{
			NUInt width, height;
			NURational frameRate;
			NVideoInterlaceMode interlaceMode;
			NURational aspectRatio;

			result = NVideoFormatGetWidth(hFormat, &width);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get width for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetHeight(hFormat, &height);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get height for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetFrameRate(hFormat, &frameRate);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get frame rate for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetInterlaceMode(hFormat, &interlaceMode);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get interlace mode for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetPixelAspectRatio(hFormat, &aspectRatio);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get aspect ratio for specified format (result = %d)!\n"), result);
				return;
			}
			printf(N_T("video format .. %dx%d @ %d/%d (interlace: %d, aspect ratio: %d/%d)\n"), width, height,
				frameRate.Numerator, frameRate.Denominator, interlaceMode, aspectRatio.Numerator, aspectRatio.Denominator);
		}
		break;
	case nmtAudio:
		{
			NInt channelCount, sampleRate, bitsPerChannel;

			result = NAudioFormatGetChannelCount(hFormat, &channelCount);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get channel count for specified format (result = %d)!\n"), result);
				return;
			}
			result = NAudioFormatGetSampleRate(hFormat, &sampleRate);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get sample rate for specified format (result = %d)!\n"), result);
				return;
			}
			result = NAudioFormatGetBitsPerChannel(hFormat, &bitsPerChannel);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get bits per sample for specified format (result = %d)!\n"), result);
				return;
			}
			printf(N_T("audio format .. channels: %d, samples/second: %d, bits/channel: %d\n"),
					channelCount, sampleRate, bitsPerChannel);
		}
		break;
	default:
		printf(N_T("unknown media type specified in format!\n"));
		return;
	}
}

static NResult ReadSoundBuffers(HNMediaReader hReader, NInt bufferCount)
{
	HNSoundBuffer hSoundBuffer = NULL;
	NResult result;
	HNMediaFormat * arFormats = NULL;
	HNMediaFormat hCurrentFormat = NULL;
	NLong length;
	NInt formatCount = 0, j;
	HNMediaSource hSource = NULL;
	NBool finishCapture = NFalse;

	result = NMediaReaderGetSource(hReader, &hSource);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media source from media reader (result = %d)!\n"), result);
		goto FINALLY;
	}

	result = NMediaReaderGetLength(hReader, &length);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media length from media reader (result = %d)!\n"), result);
		goto FINALLY;
	}
	printf(N_T("media length: %lld\n"), length);

	result = NMediaSourceGetFormats(hSource, nmtAudio, &arFormats, &formatCount);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media formats from media source (result = %d)!\n"), result);
		goto FINALLY;
	}
	printf(N_T("format count: %d\n"), formatCount);
	for (j = 0; j < formatCount; j++)
	{
		printf(N_T("[%d] "), j);
		DumpMediaFormat(arFormats[j]);
	}

	result = NMediaSourceGetCurrentFormat(hSource, nmtAudio, &hCurrentFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media formats from media source (result = %d)!\n"), result);
		goto FINALLY;
	}
	if (hCurrentFormat)
	{
		printf(N_T("current media format:\n"));
		DumpMediaFormat(hCurrentFormat);

		printf(N_T("set the last supported format (optional) ... "));
		result = NMediaSourceSetCurrentFormat(hSource, nmtAudio, arFormats[formatCount - 1]);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get media formats from media source (result = %d)!\n"), result);
			goto FINALLY;
		}
		printf(N_T("format setted\n"));
	}
	else
	{
		printf(N_T("current media format is not yet available (will be availble after media reader start)\n"));
	}

	result = NObjectSet(NULL, &hCurrentFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to clear format (result = %d)!\n"), result);
		goto FINALLY;
	}

	printf(N_T("starting capture ... "));
	result = NMediaReaderStart(hReader);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to start media reader (result = %d)!\n"), result);
		goto FINALLY;
	}
	finishCapture = NTrue;
	printf(N_T("capture started\n"));

	result = NMediaSourceGetCurrentFormat(hSource, nmtAudio, &hCurrentFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get current format (result = %d)!\n"), result);
		goto FINALLY;
	}
	if (!hCurrentFormat)
	{
		printf(N_T("current media format is not set even after media reader start!\n"));
		goto FINALLY;
	}
	DumpMediaFormat(hCurrentFormat);

	for (j = 0; j < bufferCount; j++)
	{
		NLong timespan = -1, duration = 0;
		NInt sampleRate, length;
		
		result = NMediaReaderReadAudioSample(hReader, &timespan, &duration, &hSoundBuffer);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to read audio sample from media reader (result = %d)!\n"), result);
			goto FINALLY;
		}
		if (!hSoundBuffer)
		{
			break;
		}

		result = NSoundBufferGetSampleRate(hSoundBuffer, &sampleRate);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get sound buffer sample rate (result = %d)!\n"), result);
			goto FINALLY;
		}

		result = NSoundBufferGetLength(hSoundBuffer, &length);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get sound buffer sample length (result = %d)!\n"), result);
			goto FINALLY;
		}

		printf(N_T("[%llu %llu] sample rate: %d, sample length: %d\n"), timespan, duration, sampleRate, length);
		
		result = NObjectSet(NULL, &hSoundBuffer);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to clear sound buffer (result = %d)!\n"), result);
			goto FINALLY;
		}
		fflush(stdout);
	}
	result = N_OK;
FINALLY:
	if (finishCapture)
	{
		result = NMediaReaderStop(hReader);
		if (NFailed(result)) PrintErrorMsg(N_T("failed to stop media reader (result = %d)!\n"), result);
	}
	{
		NResult result2;
		result2 = NObjectSet(NULL, &hSoundBuffer);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear sound buffer (result = %d)!\n"), result2);
		result2 = NObjectSet(NULL, &hCurrentFormat);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear current format (result = %d)!\n"), result2);
		result2 = NObjectUnrefArray(arFormats, formatCount);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear formats (result = %d)!\n"), result2);
		result2 = NObjectSet(NULL, &hSource);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear source (result = %d)!\n"), result2);
	}
	return result;
}
